sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala

```diff
@@ -645,6 +646,34 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
         },
         errorClass = "COLLATION_MISMATCH.IMPLICIT"
       )
+
+      // check if substring passes through implicit collation
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql(s"SELECT substr('a' COLLATE UNICODE, 0, 1) == substr('b' COLLATE UNICODE_CI, 0, 1)")
+        },
+        errorClass = "COLLATION_MISMATCH.IMPLICIT"
+      )
+
+      checkAnswer(spark.sql("SELECT collation(:var1 || :var2)",
+        Map(
+          "var1" -> Literal.create("a", StringType(1)),
+          "var2" -> Literal.create("b", StringType(2))
+      )
+      ),
+        Seq(Row("UTF8_BINARY"))
+      )
+
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> "UNICODE") {
+        checkAnswer(spark.sql("SELECT collation(:var1 || :var2)",
+          Map(
+            "var1" -> Literal.create("a", StringType(1)),
+            "var2" -> Literal.create("b", StringType(2))
+          )
+        ),
+          Seq(Row("UNICODE"))
+        )
+      }
```

Comment:
@srielau Is this the expected behaviour? Apparently we can pass different collations to parameters. I understood that the behaviour should be if StringType has priority of default then it has to have session level default collation, as otherwise we might have different collations with same default priority, which is not covered by design.

Reply:
Why I ask this is because by implementation we are using Parameter which binds an Expression to a name/postition depending on the query type. Allowed expression can be CreateArray, which produces implicit collated elementType, should we cast it to default if the value comes from a parameter. The thing is that if I only cast the flag to defaultST, I might get two arrays of different collationId, but same priority of DafaultST. Current implementation casts any StringType or ArrayType(StringType) to session level default StringType/ ArrayType with that session level default StringType.

Reply:
I don't know your internal design (and I shouldn't need to), so I won't comment on that.
But these examples seem correct. What is this 1/2 business in StringType() about?
Basically the user has two choices:
* say nothing: parameter marker goes with DEFAULT and looses against columns
* explicitly collate: parameter marker goes with explicit collation and wins against columns.

And if you re-use the same marker in multiple places you can choose differently.